### ISSUE TITLE
Failed to register: [413] Rate limit exceeded: 413

### ISSUE THREAD

**Issue by regan-a** [AsamK/signal-cli#1202]
I have been trying to register my land line phone number but every time I try I get the following error:

Failed to register: [413] Rate limit exceeded: 413\n

I understand that there is a limit to the number of times you are allowed to attempt a registration over a 24-hour period but I'm getting this response even if I wait a week and then submit another registration request. I'm wondering why this is happening and if there are any other reasons why I'd receive a "Rate limit exceeded" message besides hitting an actual rate limit (which in my case, seems to be 0)?

**Comment by mattwr18** [comment]
@regan-a I am also experiencing this same issue this week. I haven been trying to register a landline as well and waited two days and still getting this rate limit exceeded error on the first attempt to register.

`signal-cli v0.11.4`

**Comment by ddawson** [comment]
Same problem here. Have tried several times, including twice each after waiting over 24 hours, and most recently after generating a captcha token first, so I ran the command for the first time in over 24 hours with the token.

I have another number that can receive SMS that I tried. No problem there. The issue only seems to occur when attempting voice verification.

**Comment by frozen-sky** [comment]
I did some more research. In my case it seems the land-line numbers are fully blocked by signal. I discovered in an attempt to circumvent this issue, I tried to install the official Signal app on a (virtual) android device (with the idea to copy the keys to signal-cli). 
It literally looks like it blocks every land line number in my country (Netherlands/+31), with the very non-descriptive "Network connection error" message (dried different source IPs and different phone numbers). The debug log showed "Bad Gateway" errors. Mobile phone numbers do work. Both in signal-cli and signal app. I tried mobile phone with voice call validation (from app), it was allowed to do, but the call never arrived. 

Looks some changes at Signal side of things, that said, the "Bad Gateway" errors / Non-descriptive errors do point there are still some missing error handling or something. Hopefully this will be restored...

**Comment by ddawson** [comment]
> If I feel brave enough I will try to hack a session id in signal-cli...

Do let us know if you come up with something usable. Thanks.

**Comment by AsamK** [comment]
Thanks for the analysis, I've started working on updating libsignal-service-java which brings the API with the session id, but haven't finished that yet.

**Comment by frozen-sky** [comment]
I am not 100% if that will work. With land-lines/non-SMS lines, the twilio API (API used by Signal), throws an error. I am not 100% convinced if voice is possible in that session later again. Once again, i am not a java expert, so I had a hard time understanding the code.

**Comment by AsamK** [comment]
I've pushed a commit that implements the session based registration. I had success requesting a voice verification, by first requesting an sms and a minute later requesting the voice verification.
Please try the development build if it works for you now:
https://github.com/AsamK/signal-cli/actions/runs/4582714390

**Comment by ddawson** [comment]
So, just to be clear, the SMS part is required?

**Comment by AsamK** [comment]
Looks like it, I get a `Rate limit exceeded: 429` error without it.

**Comment by ddawson** [comment]
Okay, I finally had an opportunity to try this. It's still not working for me. When I try the SMS method, it gives `Bad response: 502`. The voice verification then fails with 429. If there's something I'm doing wrong, I really can't be sure.

* Does it matter if I first generate a CAPTCHA token and register with it, and not try without one?
* Does it matter if I reuse the token when doing the voice step?
* Does it matter if signal-cli still has data stored for the not-yet-registered account from previous attempts at this, i.e. should I delete what it has when I try again?

Maybe these don't matter. But there are enough moving parts, I'm not sure.

I wouldn't usually ask, but in this case, it would be nice if there were some step-by-step instructions telling **exactly** what to do and what not to do. Or at least exactly what *you* did.

**Comment by AsamK** [comment]
I tried it with a new account with a captcha the first time.
I don't know how it works exactly, someone needs to try that out.
I'd suggest first deleting the existing account data and trying again.

**Comment by frozen-sky** [comment]
I tried as well, and got the same behaviour as ddawson. 

According to [comment](https://github.com/signalapp/Signal-Android/issues/12839#issuecomment-1481168774) i quoted before, it looks like the "502 Bad Gateway" is a result from an error code received from Signal's Phone verifier. If I am not mistaken, this is twilio. I've created an account myself at twilio, and indeed, if I try to verify with a land-line, it immediately throws an error. Voice still works there. I think the registration backend of Signal, does not start the whole registration session if the backend throws an error, and thus, you can not finish the registration with voice after one-two minutes. So my best guess is, if twilio does not recognize your phone number as text-service-ready, there is no way to register with a land-line. 

Guess have to ask the Signal folks if that will be their policy and/or if that is a bug they want to solve eventually.

**Comment by ddawson** [comment]
Also, if signal-cli is unable to do this, is there any way to do the registration on the app or something else, and then fix things so that signal-cli becomes the primary device? I know it's not as simple as just copying data over, since they almost certainly store the data in different ways. But maybe there's something.

**Comment by frozen-sky** [comment]
As far as my research went: No. 
I read the code of the registration service of Signal, and if your phone number can not receive short code text messages (you can test with Twilio), you won't be able to register. Not with app, not with signal-cli, not with anything.

I want to raise this with the Signal Team to know if this is a (new?) policy from them or a bug/side effect of the changes they did. But didn't find the time yet to do that.

**Comment by frozen-sky** [comment]
@csouti-chaps, do you know if you phone number can receive short code SMS (even though they may not arrive?). As far as i understand the recent test build of signal-cli, in that case it should be able to register via first SMS, wait and then Voice.

**Comment by csouti-chaps** [comment]
@frozen-sky yes I could register my land line, by voice and exclusely with signal-cli like that : 

1. register by sms (without --voice param, even if can't receive sms)
2. wait 60 seconds (like android app)
3. register by voide (with --voice, I received phone call to give me verifcation code)
4. verify...

My phone number is a land line, normally my operator (free in france) don't provide sms support.
@ddawson with this workflow you should be able to register your line with signal-cli as primary device, no ?

**Comment by csouti-chaps** [comment]
I provided one captcha for step 1 and second captcha for step 3 with this method https://github.com/AsamK/signal-cli/wiki/Registration-with-captcha

**Comment by ddawson** [comment]
Nope. Still not working for me. I still get a 502 when trying the SMS registration.

**Comment by etlweather** [comment]
> @frozen-sky yes I could register my land line, by voice and exclusely with signal-cli like that :
> 
> 1. register by sms (without --voice param, even if can't receive sms)
> 2. wait 60 seconds (like android app)
> 3. register by voide (with --voice, I received phone call to give me verifcation code)
> 4. verify...
> 
> My phone number is a land line, normally my operator (free in france) don't provide sms support. @ddawson with this workflow you should be able to register your line with signal-cli as primary device, no ?

Worked for me, with v0.11.8 (not sure that version matter, based on release notes).

I did like written above - registered my landline without `--voice`, waiting a tad over 60 sec and then did with the `--voice`.

**Comment by nblock** [comment]
> Nope. Still not working for me. I still get a 502 when trying the SMS registration. There seems to be some difference depending on country or provider.

I do have the same issue - without `--voice`, I get a 502 response code and with `--voice`, a `429 Rate limit exceeded`. I tried multiple landline numbers but none of them seem to work.

**Comment by etlweather** [comment]
You have to also use captcha. Additionally, it does not take long before you are throttled if you don't follow the steps exactly. A few tries and bang, you're on the throttle list and you have to wait quite some time before it's freed up to use again.

**Comment by ddawson** [comment]
> You have to also use captcha.

Please read my above comments. I did consider that, and used them. It didn't help.

> A few tries and bang, you're on the throttle list and you have to wait quite some time before it's freed up to use again.

How long? Days? Or even longer? Because I did wait days and made sure to get everything right, but it still didn't help.

Also, even if it's as you say, the system is too unforgiving. Something needs to change. If not how many tries are allowed, then there should be more of a warning _somewhere_.

**Comment by etlweather** [comment]
> How long? Days? Or even longer? Because I did wait days and made sure to get everything right, but it still didn't help.

When I got rate limtted on my cell phone (not even signal-cli) in the past due to not remembering my PIN, I was locked for a week.

> Also, even if it's as you say, the system is too unforgiving. Something needs to change. If not how many tries are allowed, then there should be more of a warning somewhere.

Yes, better communication from Signal would be nice. But please understand that this is something on *Signal* side, not `signal-cli`. `signal-cli` is just a client connecting to Signal server and does not control how Signal server works.

**Comment by ddawson** [comment]
Yes, I understand that already. But I thought that needed to be said.

**Comment by theblackhole** [comment]
> @frozen-sky yes I could register my land line, by voice and exclusely with signal-cli like that :
> 
> 1. register by sms (without --voice param, even if can't receive sms)
> 2. wait 60 seconds (like android app)
> 3. register by voide (with --voice, I received phone call to give me verifcation code)
> 4. verify...
> 
> My phone number is a land line, normally my operator (free in france) don't provide sms support. @ddawson with this workflow you should be able to register your line with signal-cli as primary device, no ?

It seems that it follows the same verification procedure as that of the application (we must wait before asking for a verification call). It worked with my land line ("Free" ISP in France too), with two freshly generated captchas for each registration attempt (the first one without `--voice` and the second with `--voice`).

I [updated the Wiki](https://github.com/AsamK/signal-cli/wiki/Quickstart/_compare/f33d750f7a9f7ce3347a1bf5ece1c8c97131ddaf...21713ccaf0d6e4e5e30a488cdfe5c5cb40fdf3b7) to reflect these changes in the verification process.

I think that the first registration attempt (with SMS) needs to "succeed" (=no reply, no error) to allow a later attempt with `--voice`. Following this logic, it seems normal to me to get a "429 Rate Limit Exceeded" if the first step gave an error.

Also from what I could see about the "Rate limit exceeded", it seems that there would be a difference between the 413 code and 429 one:
* `[413] Rate limit exceeded: 413` seems to indicate a real limit where you have to wait before trying again (usually at least 24h, I don't know if this is the case with Signal since OP tried after a week and it gave the same error)
* `[429] Rate limit exceeded: 429` seems to only appear for verification with `--voice` and therefore does not seem to indicate a real restriction since it is possible to register immediatly after by following the two-step procedure above.

**Comment by mh-g** [comment]
I have tried several landline numbers (in Germany). The sequence that worked for me was:
`signal-cli -u +49XXX register`
Create a captcha and use it:
```
signal-cli -u +49XXX register --captcha signal-hcaptcha….
signal-cli -u +49XXX register --voice
signal-cli -a +49XXX verify
```
(In fact, I cheated a bit and used a landline number appended by a single digit to be able to have several signal instances registered for several servers. Some of the digits got "burned" with 429 errors and waiting for >36 hours did NOT help.)

**Comment by lindes** [comment]
I'm not sure if this is sufficiently related to be worth noting here, but since it's an open issue with a similar error message, I'm starting here.  I can log a separate issue, and/or upgrade and try again before further stuff, but I thought I'd start by mentioning here that I'm getting the same 413 Rate limit exceeded when trying to do a different command:

    signal-cli -o json receive

This gets a bunch of messages received just fine, and then finally ends with:

    WARN  SendHelper - Sending failed due to rate limiting from the signal server: [413] Rate limit exceeded: 413

This from a built-from-git version wherein `signal-cli --version` returns `signal-cli 0.11.6` (built from commit 161ec9f8).

Please let me know if I should file a separate issue, and/or update and re-build first.  (Though actually, noticing the very next commit after the one above, I'm definitely going to do an update and see if that helps things... going ahead and commenting anyway, just in case it's somehow useful to someone.  Apologies if it's noise.)

**Comment by AsamK** [comment]
rate limit when sending messages is another issue.
